libimage/push.go

```diff
@@ -24,33 +30,93 @@ type PushOptions struct {
 // Return storage.ErrImageUnknown if source could not be found in the local
 // containers storage.
+// Returns the bytes of the copied manifest when pushing a single tag,
+// which may be used for digest computation.
+// When pushing with AllTags=true then the returned []byte is always nil.
 func (r *Runtime) Push(ctx context.Context, source, destination string, options *PushOptions) ([]byte, error) {
 	if options == nil {
 		options = &PushOptions{}
 	}
 
-	// Look up the local image.  Note that we need to ignore the platform
-	// and push what the user specified (containers/podman/issues/10344).
-	image, resolvedSource, err := r.LookupImage(source, nil)
+	// Push the single image
+	if !options.AllTags {
+
+		// Look up the local image.  Note that we need to ignore the platform
+		// and push what the user specified (containers/podman/issues/10344).
+		image, resolvedSource, err := r.LookupImage(source, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Make sure we have a proper destination, and parse it into an image
+		// reference for copying.
+		if destination == "" {
+			// Doing an ID check here is tempting but false positives (due
+			// to a short partial IDs) are more painful than false
+			// negatives.
+			destination = resolvedSource
+		}
+
+		return pushImage(ctx, image, destination, options, resolvedSource, r)
+	}
+
+	// Below handles the AllTags option, for which we have to build a list of
+	// all the local images that match the provided repository and then push them.
+	//
+	// For now, make sure a destination was not specified and get it from the source.
+	// This could change in the future, but that gets close to the Copy() functionality.
+	if len(destination) != 0 {
+		return nil, fmt.Errorf("`destination` should not be specified if using AllTags")
+	}
+
+	// Make sure the source repository does not have a tag
+	srcNamed, err := reference.ParseNormalizedNamed(source)
 	if err != nil {
 		return nil, err
 	}
+	if !reference.IsNameOnly(srcNamed) {
+		return nil, fmt.Errorf("can't push with AllTags if source tag is specified")
+	}
+
+	logrus.Debugf("Finding all images for source %s", srcNamed.Name())
+	listOptions := &ListImagesOptions{}
+	srcImages, _ := r.ListImages(ctx, []string{srcNamed.Name()}, listOptions)
 
+	// Push each tag for every image in the list
+	for _, img := range srcImages {
+		namedTagged, err := img.NamedTaggedRepoTags()
+		if err != nil {
+			return nil, err
+		}
+		for _, n := range namedTagged {
+			// Filter on repo name again to avoid pushing an image that matches
+			// the source image ID but has a different repository than the source
+			currentNamed, err := reference.ParseNormalizedNamed(n.Name())
+			if err != nil {
+				return nil, err
+			}
+			if reference.Path(currentNamed) == reference.Path(srcNamed) {
+				// Have to use Sprintf because pushImage expects a string
+				destWithTag := fmt.Sprintf("%s:%s", source, n.Tag())
+				_, err := pushImage(ctx, img, destWithTag, options, "", r)
```

Comment:
Conceptually I’d prefer to move the `destRef := ParseImageName` etc. logic from `pushImage` into the `!AllTags` case above, and have `pushImage` accept a `types.ImageReference`; then this can use `docker.NewReference` instead of making another round-trip via an untyped string.

But I haven’t really examined how that would work with the `writeEvent` use of `destination`, so I’m not completely sure it would be viable.

---

Why is this combining `source` and `n` at all? We know the repo of `currentNamed` and `srcNamed` matches, the two only differ in the tag; so even if this had to use a string, we could use `n.String()`, couldn’t we?